src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreOperationLogger.cs

```diff
@@ -135,106 +113,117 @@ public Task StopAsync()
 
-        public override void LogError(string data)
+        /// <summary>
+        /// Log messages to VS. This is optimized to determine
+        /// if the message needs to be logged before moving to the
+        /// UI thread. 
+        /// </summary>
+        public override Task LogAsync(ILogMessage logMessage)
         {
-            LogToVS(VerbosityLevel.Quiet, data);
-        }
+            HandleErrorsAndWarnings(logMessage);
 
-        public override void LogInformationSummary(string data)
-        {
-            // Treat Summary as Debug
-            LogDebug(data);
+            if (ShouldShowMessage(logMessage))
+            {
+                var verbosityLevel = GetMSBuildLevel(logMessage.Level);
+                var reportProgress = ShouldReportProgress(logMessage);
+
+                // Write to the output window if the verbosity level is high enough.
+                var showAsOutputMessage = ShouldShowMessageAsOutput(verbosityLevel);
+
+                // Avoid moving to the UI thread unless there is work to do
+                if (reportProgress || showAsOutputMessage)
+                {
+                    // Run on the UI thread
+                    return DoAsync((_, progress) =>
+                    {
+                        LogToVS(reportProgress, showAsOutputMessage, logMessage, progress);
+                    });
+                }
+            }
+
+            return Task.FromResult(0);
         }
 
-        private void LogToVS(VerbosityLevel verbosityLevel, string message)
+        private void LogToVS(bool reportProgress, bool showAsOutputMessage, ILogMessage logMessage, RestoreOperationProgressUI progress)
         {
-            if (_cancelled)
+            var verbosityLevel = GetMSBuildLevel(logMessage.Level);
+
+            // Progress dialog
+            if (reportProgress)
             {
-                // If an operation is canceled, don't log anything, simply return
-                // And, show a single message gets shown in the summary that package restore has been canceled
-                // Do not report it as separate errors
-                return;
+                progress?.ReportProgress(logMessage.Message);
             }
 
-            // If the verbosity level of message is worse than VerbosityLevel.Normal, that is,
-            // VerbosityLevel.Detailed or VerbosityLevel.Diagnostic, AND,
-            // _msBuildOutputVerbosity is lesser than verbosityLevel; do nothing
-            if (verbosityLevel > VerbosityLevel.Normal && OutputVerbosity < (int)verbosityLevel)
+            // Output console
+            if (showAsOutputMessage)
             {
-                return;
+                WriteLine(verbosityLevel, logMessage.Message);
             }
+        }
+
+        private void HandleErrorsAndWarnings(ILogMessage logMessage)
+        {
+            var verbosityLevel = GetMSBuildLevel(logMessage.Level);
 
             // VerbosityLevel.Quiet corresponds to ILogger.LogError, and,
```

Comment:
`LogLevel.Error` and `LogLevel.Warning`. `ILogger `does not contain verbosity levels.